WIP - Add support reading SC3ML v0.10; ResponseIIR & decimation on PAZ

Jollyfant opened the pull request:
This is a work in progress. Added code to read SC3ML version 0.10; also refactored some of the code.. for discussion see issue #2004

megies commented:
@Jollyfant this doesn't look right and I think the reason is #2130. It looks like you were trying to rebase? If so, I would recommend to simply start from current master again and instead just cherry-pick that dozen commits manually. For me this used to be the simpler solution the last couple of times, as git was getting real confused during rebasing.

Jollyfant commented:
Yeah I boned it.. I will redo it on current master

Jollyfant commented:
Seems to be rebased on current master now 🎉 but I need to review this some time this week.. when are you pushing for 1.2.0?

megies commented:
> Seems to be rebased on current master now tada but I need to review this some time this week.. when are you pushing for 1.2.0?

There's still some time I think, a couple of weeks maybe. Dunno. Or longer if nobody pushes.

Jollyfant commented:
Gotcha but meanwhile CI seems to be tripping up on something

megies commented:
That fails unrelated, comes from a new numpy choking up on sth in our code.

megies commented:
Can you rebase on current master? Should get rid of unrelated fails..

megies commented:
@Jollyfant as mentioned by Gempa staff on our users mailing list, there's also the `responseFAP` tag that needs to be implemented.

megies commented:
@Jollyfant I added the responseFAP element and also properly made sensor/datalogger optional on channels.

I'll leave the rest to you, since you're most familiar with this submodule.

megies commented:
@Jollyfant I felt free to rebase on current master.. so please update your local branches

Jollyfant commented:
Thanks @megies. I will put this on the to-do list and have a look at some clean-up for this module too. It could use some attention..

Jollyfant commented:
I've included the code by @QuLogic from #2260 for better pole & zero parsing here.

megies commented:
> I've included the code by @QuLogic from #2260 for better pole & zero parsing here.

@Jollyfant you do realize you're hijacking a PR of a new contributor there? That's kinda rude.. robbing new :sparkles:  contributors of their commits :stuck_out_tongue_winking_eye:  I'm totally fine with merging these PRs, if that makes it easier for you to continue work on this one, but you should have cherry-picked all of @thefroid's commits from #2260 first before continuing with adapting to @QuLogic's regex improvements.. :wink:

megies commented:
@Jollyfant I finalized #2260 and once that is merged into maintenance and master, you should ideally get rid of the latest two commits in this PR and rebase on master.

megies commented:
@Jollyfant, #2260 is in master now.

megies commented:
Rebased on current master and force-pushed so that we have fresh CI results tomorrow.

Jollyfant commented:
Ok I will put this on the list for next week for some final fixes and touch ups. There is also this issue #2296 that would be nice to include in the next release.

Jollyfant commented:
What the heck.. what's with all these extra files? These are the only files changed:

    On branch add-sc3ml-010
    Changes to be committed:
      (use "git reset HEAD <file>..." to unstage)

	modified:   CHANGELOG.txt
	modified:   obspy/io/seiscomp/inventory.py
	new file:    obspy/io/seiscomp/test/data/version0.12
	modified:  obspy/io/seiscomp/test/test_inventory.py

Anyway.. I cherry picked the commits from @medlin01GA and put them in this pull request.

Jollyfant commented:
@megies is that some GitHub bug? Because most of these files show no changes or -1 bytes..

megies commented:
In any case it's this "fetch master" commit of yours: b9dbce72a41847020c6e7d640e71e535117750e8

Jollyfant commented:
Can it be a problem with line endings?

megies commented:
Could be, have a look at #1978 but even then you can still not commit those local changes in all files.. I always use `git add -p` to stage commits.

Jollyfant commented:
Idk how to go back and undo this mess. Probably better to clone master and copy the changes in again?

megies commented:
You should be able to go back to https://github.com/obspy/obspy/commit/c2a4d02d8aceb8c7f1dd8b703816093d79bf9b97 and try again, I think

![screenshot from 2019-02-22 12-31-20](https://user-images.githubusercontent.com/1842780/53240039-dbfbfa00-369d-11e9-9864-47b0d83800ca.png)

krischer commented:
Something went really wrong with the history here. I tried to fix it to the best of my ability and cleaned up the code a bit so now at least superficially it looks alright. I know little about SC3ML so someone will need to have a look. This is a fairly big change and IMHO there are not enough new tests for such a change.

@Brtle @Jollyfant @medlin01GA  can you have a look at the current stage and make sure this is correct and all necessary things are tested? Also to be it is a bit unclear which SC3ML versions are now supported. Everything up to 0.12?

A backup branch before my rebasing can be found here but its history is pretty much unreadable: https://github.com/krischer/obspy/tree/add-sc3ml-010

medlin01GA commented:
There are additional fixes that think should be brought into master: 
* Performance fix in `io/seiscomp/inventory.py`, function `_tag2obj`: https://github.com/medlin01GA/obspy/commit/d2829561dae9ff25e6afe57a0905db353539916b
* Path comparison fix in `core/util/misc.py` that affects tests under Windows: https://github.com/medlin01GA/obspy/commit/813a722dc71a2617915f8b4684e1a08f9f720f4c
* Unit tests fix in `io/seiscomp/tests/test_inventory.py` where warnings filter could be incorrect (seemed to depend on order of test execution): https://github.com/medlin01GA/obspy/commit/c78661274eb3a14b808f0950768c94104326bfab
* Unit test fixes for Windows 64 bit in `core/tests/test_stream.py`, where small floating point differences could affect tests: https://github.com/medlin01GA/obspy/commit/a2729dd19e0560c741240af05cbcc5d09786fd24 and https://github.com/medlin01GA/obspy/commit/bb42892cf51ada7efd9095f5ebd733aedb5f4c27
* OPTIONAL - made test file `io/seiscomp/tests/data/EB_response_sc3ml` more human-readable: https://github.com/medlin01GA/obspy/commit/f8e471e4bcc1fb06f6e497b8260435ff8801c020

Jollyfant commented:
@medlin01GA 
> I have visually reviewed the cherry picked integrations from obspy#2296 (thank you @Jollyfant) and it all looks fine. I note that the responses sub-dictionary of the instrument register has been flattened to remove the response type from the index. I don't have enough expertise in sc3ml format to be sure about the safety of doing this, since it assumes the response ID will never be duplicated in different response types. However if this assumption is valid, then it's a nice simplification.

I did this because the `publicID` does not necessarily need to be prefixed with the response type (e.g. https://github.com/obspy/obspy/blob/master/obspy/io/seiscomp/tests/data/sc3ml_empty_depth_and_id.sc3ml). Then it doesn't know which dictionary to look in so I flattened it.

@krischer Thanks for the effort. Schema versions 0.7 - 0.11 are supported after this PR.

Brtle commented on CHANGELOG.txt:
```diff
@@ -78,7 +78,9 @@ master:
  - obspy.io.seiscomp:
-   * Adding support for SC3ML 0.10 (see #2024).
+   * Adding support for SC3ML 0.10 (see #2004, #2024 and #2124).
+   * Adding support for SC3ML 0.11 (see #2124).
```
Currently the event plugin only supports version 0.10. I did a PR (#2284) to support 0.11 but it isn't merged. The changelog is a bit misleading IMHO.

Brtle commented on obspy/io/seiscomp/inventory.py:
```diff
@@ -82,7 +85,7 @@ def _read_sc3ml(path_or_file_object):
     root = etree.parse(path_or_file_object).getroot()
 
-    # Code can be used for version 0.7, 0.8, and 0.9
+    # Code can be used for version 0.7, 0.8, 0.9, and 0.10
```
Also version 0.11

Brtle commented on obspy/io/seiscomp/tests/test_inventory.py:
```diff
@@ -140,6 +139,48 @@ def test_compare_upper_level(self):
+        Test reading SC3ML schema version 0.10
+        """
+        data_file = os.path.join(self.data_dir, "sc3ml_010.xml")
+        with warnings.catch_warnings(record=True) as w:
```
The unit tests `test_empty_detph` and `test_sc3ml_v010` fail locally when I run the test with :

```bash
obspy-runtests obspy.io.seiscomp.tests.test_inventory
```

but it works with only `obspy-runtests`. Maybe add a `warnings.simplefilter("always")` ?

```python
with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
```

krischer commented:
Does anyone of you have time to do this soonish? I don't have the expertise here but it would be an important feature to have in the next release.

Jollyfant commented:
> Does anyone of you have time to do this soonish? I don't have the expertise here but it would be an important feature to have in the next release.

I think all we need is some extra tests. I'm looking through inventories to find an IIR response type in the wild but so far no luck.

medlin01GA commented on obspy/io/seiscomp/inventory.py:
```diff
@@ -366,34 +417,32 @@ def _read_channel(inventory_root, cha_element, _ns):
+            if len(response_id.split("#")) == 1:
+                resp_type = response_id.split("/")[0]
+            else:
+                resp_type = response_id.split("#")[0]
```
Something doesn't look right here - the variable `resp_type` appears to never be used.

megies commented:
@Jollyfant I just started from master and cherry-picked all the older commits, pushed the results here:  https://github.com/megies/obspy/commits/clean_2124

That should only lack @medlin01GA's commits now, I am not sure which ones they are, tbh. Your old commit message 'Cherry pick commits by @medlin01GA; support 0.11; minor fixes - WIP' doesn't look like a cherry pick really, did you edit in the changes back then by hand? Hope it helps, in any case.. that's why I usually avoid merging together two separate PR efforts working on the same code on the fly. One PR should be properly merged first and then the other PR has to adapt to the finalized newly changed master.

medlin01GA commented:
@Tobias Megies<mailto:notifications@github.com>, I'm a little concerned about the changes for #2124. I noticed something yesterday and commented here: https://github.com/obspy/obspy/pull/2124#discussion_r285552893 . It would be better if my query were addressed before completing the integrations.


Also, I have almost got a clean PR based off recent origin/master with my changes from #2296. The main reason I didn't already push it out for review is that I knew it would conflict with #2124.


I think it would be better if #2124 did not have any of my changes (which were a manual integration more than a technical "cherry pick") and stuck to just the changes it originally intended. Then these two PRs would be properly separated, and could be committed to master in orderly sequence.  My changes really need to be reviewed independently of the changes in #2124, as there could be revisions to the implementation details coming out of a disciplined process.



________________________________
From: Tobias Megies <notifications@github.com>
Sent: Wednesday, 22 May 2019 7:14:18 PM
To: obspy/obspy
Cc: Medlin Andrew; Mention
Subject: Re: [obspy/obspy] WIP - Add support reading SC3ML v0.10; ResponseIIR & decimation on PAZ (#2124)


@Jollyfant<https://github.com/Jollyfant> I just started from master and cherry-picked all the older commits, pushed the results here: https://github.com/megies/obspy/commits/clean_2124

That should only lack @medlin01GA<https://github.com/medlin01GA>'s commits now, I am not sure which ones they are, tbh. Your old commit message 'Cherry pick commits by @medlin01GA<https://github.com/medlin01GA>; support 0.11; minor fixes - WIP' doesn't look like a cherry pick really, did you edit in the changes back then by hand? Hope it helps, in any case.. that's why I usually avoid merging together two separate PR efforts working on the same code on the fly. One PR should be properly merged first and then the other PR has to adapt to the finalized newly changed master.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/obspy/obspy/pull/2124?email_source=notifications&email_token=AK3ZLRYW2CWTEBUEBPTW6SLPWUFGVA5CNFSM4E4T6VWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV6NWTQ#issuecomment-494721870>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AK3ZLR6VBW77V4MKBN52LYTPWUFGVANCNFSM4E4T6VWA>.

Geoscience Australia Disclaimer: This e-mail (and files transmitted with it) is
intended only for the person or entity to which it is addressed. If you are not
the intended recipient, then you have received this e-mail by mistake and any
use, dissemination, forwarding, printing or copying of this e-mail and its file
attachments is prohibited. The security of emails transmitted cannot be
guaranteed; by forwarding or replying to this email, you acknowledge and accept
these risks.
-------------------------------------------------------------------------------------------------------------------------

megies commented:
> Also, I have almost got a clean PR based off recent origin/master with my changes from #2296. The main reason I didn't already push it out for review is that I knew it would conflict with #2124. I think it would be better if #2124 did not have any of my changes (which were a manual integration more than a technical "cherry pick") and stuck to just the changes it originally intended. Then these two PRs would be properly separated, and could be committed to master in orderly sequence.

Sure, I'm all for clean and separate PRs. You and @Jollyfant just need to agree on the way forward and battle your way through. Was just trying to help out.

medlin01GA commented:
@Jollyfant<mailto:notifications@github.com> if you can excise the changes in your PR from #2296, then I'm happy for you to go ahead and re-issue your PR. Once it was in master, I would then resolve the conflicts prior to regenerating my PR off latest master. Would that work for you?


@Tobias<mailto:notifications@github.com> I appreciate you pushing to progress and resolve this.


________________________________
From: Tobias Megies <notifications@github.com>
Sent: Wednesday, 22 May 2019 11:54:18 PM
To: obspy/obspy
Cc: Medlin Andrew; Mention
Subject: Re: [obspy/obspy] WIP - Add support reading SC3ML v0.10; ResponseIIR & decimation on PAZ (#2124)


Also, I have almost got a clean PR based off recent origin/master with my changes from #2296<https://github.com/obspy/obspy/issues/2296>. The main reason I didn't already push it out for review is that I knew it would conflict with #2124<https://github.com/obspy/obspy/pull/2124>. I think it would be better if #2124<https://github.com/obspy/obspy/pull/2124> did not have any of my changes (which were a manual integration more than a technical "cherry pick") and stuck to just the changes it originally intended. Then these two PRs would be properly separated, and could be committed to master in orderly sequence.

Sure, I'm all for clean and separate PRs. You and @Jollyfant<https://github.com/Jollyfant> just need to agree on the way forward and battle your way through. Was just trying to help out.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/obspy/obspy/pull/2124?email_source=notifications&email_token=AK3ZLR7RILGMTP3TRXOWPDLPWVGAVA5CNFSM4E4T6VWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV7D4OI#issuecomment-494812729>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AK3ZLR3KFYJLLRXH6ZX3NY3PWVGAVANCNFSM4E4T6VWA>.

Geoscience Australia Disclaimer: This e-mail (and files transmitted with it) is
intended only for the person or entity to which it is addressed. If you are not
the intended recipient, then you have received this e-mail by mistake and any
use, dissemination, forwarding, printing or copying of this e-mail and its file
attachments is prohibited. The security of emails transmitted cannot be
guaranteed; by forwarding or replying to this email, you acknowledge and accept
these risks.
-------------------------------------------------------------------------------------------------------------------------

Jollyfant commented:
@medlin01GA You mean remove the commits I picked from your branch?

medlin01GA commented:
@Jollyfant<mailto:notifications@github.com> Yes, that's right. That way the concern of each PR is kept separate, and the merging and review processes are easier. Is that manageable from your perspective?


________________________________
From: Jollyfant <notifications@github.com>
Sent: Thursday, 23 May 2019 7:25:05 PM
To: obspy/obspy
Cc: Medlin Andrew; Mention
Subject: Re: [obspy/obspy] WIP - Add support reading SC3ML v0.10; ResponseIIR & decimation on PAZ (#2124)


@medlin01GA<https://github.com/medlin01GA> You mean remove the commits I picked from your branch?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/obspy/obspy/pull/2124?email_source=notifications&email_token=AK3ZLRYCUMWHD3ZPVPAPZY3PWZPHDA5CNFSM4E4T6VWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWBTXVY#issuecomment-495139799>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AK3ZLR3YBZ4SNEPML4BUJ7DPWZPHDANCNFSM4E4T6VWA>.

Geoscience Australia Disclaimer: This e-mail (and files transmitted with it) is
intended only for the person or entity to which it is addressed. If you are not
the intended recipient, then you have received this e-mail by mistake and any
use, dissemination, forwarding, printing or copying of this e-mail and its file
attachments is prohibited. The security of emails transmitted cannot be
guaranteed; by forwarding or replying to this email, you acknowledge and accept
these risks.
-------------------------------------------------------------------------------------------------------------------------

krischer commented:
Hey guys - any change you'll get to this in the next few days? Otherwise we'd move it to the next release.

megies commented:
@Jollyfant we had a look at this but do to the mixing commits from two PRs and continuing again afterwards, we see no chance to clean this up ourselves in a reasonable amount of time or workload. Please let us know if you can clean this up in the next 1-2 days or it will probably have to be bumped from 1.2 to 2.0

Jollyfant commented:
Please bump it, I cannot get this done before 1.2.

claudiodsf commented:
Pinging @Jollyfant on this to inform that the work on the event part of SC3ML 0.11 resumed via https://github.com/obspy/obspy/pull/2284.

It would be nice to have also the inventory part done for 2.0.0 😉

claudiodsf commented:
Hi @Jollyfant!

We would like to try and include your PR in ObsPy `1.3.0` or `1.4.0`.

If you're still interested, could you please rebase against the current `master` branch and force-push? 
This will allow to rerun the tests and check what needs to be done.

Alternatively, you could give me write access to your `add-sc3ml-010` branch and I will try to do this myself 😉

Jollyfant commented:
Hi @claudiodsf, I am not certain about the state of the PR and whether it can be merged/rebased as is. Because a bunch of the code already made it in to `1.2.0`. Unfortunately I currently don't have the time to dive back in..

It is probably easier to add SC3ML v0.10 and v0.11 support by starting from current master and adding it there.

claudiodsf commented:
> It is probably easier to add SC3ML v0.10 and v0.11 support by starting from current master and adding it there.

Ok, I will then probably wait and see whether we can merge #2695 and #2284, and then try to cherry pick from this PR.

I keep this open, for the moment.
